### PR TITLE
Made Changes to Utilities.py file

### PR DESCRIPTION
**Pull request by faseehahmed26**

### CODE REVIEW

File: ivy/functional/frontends/torch/utilities.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+class Utilities:
+
+    def result_type(arg1, arg2):
+        return torch.result_type(arg1, arg2)
+
+    def can_cast(dtype1, dtype2):
+        return torch.can_cast(dtype1, dtype2)
+
+    def _assert(condition, message):
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/torch/utilities.py:15]: The same for this too

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hi @faseehahmed26 
Thanks for contributing, here are a few points to consider while implementing the frontend functions, and I recommend you to follow the contribution guideline by assigning yourself to one function at a time. [Contribution guide](https://lets-unify.ai/ivy/contributing.html)
And also you will need to implement unit tests for your implementation.

Thanks:)

**Comment by faseehahmed26** [comment]
Hi @hirwa-nshuti I have made the changes you requested could you give an update regarding my first utility function whether the unit test and function are according to the standards of ivy contribution guide

**Comment by faseehahmed26** [comment]
this issue https://github.com/unifyai/ivy/issues/18

**Comment by hirwa-nshuti** [comment]
You should follow the contribution guide to open an issue for a function you're implementing [ToDo List Issues](https://lets-unify.ai/ivy/contributing/1_the_basics.html#todo-list-issues). After that, you will comment the issue to the PR. 
Thanks

**Comment by faseehahmed26** [comment]
Ok I have one doubt can I work on this issue https://github.com/unifyai/ivy/issues/18  or is this not available for PR 
or I need to create a new issue then work on it referencing it as specified in [ToDo List Issues](https://lets-unify.ai/ivy/contributing/1_the_basics.html#todo-list-issues)

**Comment by hirwa-nshuti** [comment]
You need to create an issue with name of the function you're going to work on the comment the `- [ ] #issue` number to the main ToDo list of that function and a quick review is that I see that the function you're implementing is being implemented by some other contributor if you head to #18

File: ivy/functional/frontends/torch/utilities.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# global
+
+# local
+import ivy
+
+
+def result_type(*arg1):
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/torch/utilities.py:7]: This function is a task in progress by another contributor # #2924 As you can see from this PR would you please take another function?

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+# local
+import ivy_tests.test_ivy.helpers as helpers
+import ivy.functional.backends.numpy as ivy_np
+import ivy.functional.backends.torch as ivy_torch
+from ivy_tests.test_ivy.helpers import handle_cmd_line_args
+
+
+# result_type
+@handle_cmd_line_args
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:13]: Feel free to implement any other function from this task as this one is work in progress

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hi, @faseehahmed26 Here are a few points to note.
Thanks

**Comment by faseehahmed26** [comment]
Hi @hirwa-nshuti I am done creating can_cast function but I am unable to create a unit test for it as I am unable to understand that why its failing do you have any resources to understand hypothesis and pytest

**Comment by hirwa-nshuti** [comment]
Hi @faseehahmed26 Would you please head to some of the implemented examples from our documentation, and also please open an issue for this function you're implementing such that we can link it with this PR? [frontend guide](https://lets-unify.ai/ivy/deep_dive/16_ivy_frontends.html)

Thanks:)

**Comment by faseehahmed26** [comment]
@hirwa-nshuti  I have created a new issue of the function I am working on https://github.com/unifyai/ivy/issues/5197

**Comment by faseehahmed26** [comment]
Hi Felix I have worked  extensively on creating the unit test for can_cast function but I am unable to find a solution if possible could you please guide  me and also there is weird thing with ivy.can_cast function as its not giving the right output for a query the only possible reason I might think is it works only with ivy dtypes and not tensor
can_cast
here is the screenshot of my unit test code and the error I am getting

 
can_cast_test

can_cast_error

**Comment by hirwa-nshuti** [comment]
Hi the problem here you have is that the ivy `can_cast` function conforms to Array API standard Promotion rules if you want to test you will generate inputs bounded on these [Promotion rules](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#type-promotion)

**Comment by faseehahmed26** [comment]
can_cast_error

if you see here its checking from pytorch documentation where arguments are form,to and ivy.can_cast has arguments (from_,to)  but if I use from then its an error as from is a keyword  used to import only a specified section from a module.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# global
+import numpy as np
+from hypothesis import given, strategies as st
+
+# local
+import ivy_tests.test_ivy.helpers as helpers
+import ivy.functional.backends.numpy as ivy_np
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:7]: This line should not be imported as it handled in helpers functions

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# global
+import numpy as np
+from hypothesis import given, strategies as st
+
+# local
+import ivy_tests.test_ivy.helpers as helpers
+import ivy.functional.backends.numpy as ivy_np
+import ivy.functional.backends.torch as ivy_torch
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:8]: This as well

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+# can_cast
+@handle_cmd_line_args
+@given(
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("valid", full=True), num_arrays=1
+    ),
+    to_dtype=helpers.get_dtypes("valid", full=False),
+    num_positional_args=helpers.num_positional_args(fn_name="can_cast"),
+)
+def test_can_cast(
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:21]: the test function should be `test_torch_can_cast`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    to_dtype,
+    as_variable,
+    num_positional_args,
+    native_array,
+    container,
+    instance_method,
+    fw,
+):
+    input_dtype, x = dtype_and_x
+    helpers.test_function(
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:32]: You will need to use `helpers.test_frontend_function()`
[Test frontend function helper](https://github.com/unifyai/ivy/blob/3c5c0a866a07670d6481d42b0744c26263a4088e/ivy_tests/test_ivy/helpers/function_testing.py#L392)

File: ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        input_dtypes=input_dtype,
+        as_variable_flags=as_variable,
+        with_out=False,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        container_flags=container,
+        instance_method=instance_method,
+        fw=fw,
+        fn_name="can_cast",
+        from_=x[0],
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_utilities.py:42]: According to PyTorch docs, the input should be dtype not an array

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
The function is well implemented here are points to note in writing unit tests.
You can also try to remove unused imports.

**Comment by faseehahmed26** [comment]
Sorry I haven't uploaded the updated test function this  error I am getting is from this test file which I have uploaded right now 
which led me to think what might be the original issue is based on argument name which is from_ and whether it should be "from"  or not 
can_cast_error

**Comment by faseehahmed26** [comment]
but from is a keyword right so how should I use it as an argument

**Comment by faseehahmed26** [comment]
Hi @hirwa-nshuti I think there might be a way in ivy test where we won't require to use "from" as an argument as it is a reserved keyword I have tried various methods and reviewed various resources like 
https://bugs.python.org/issue5382 etc but can't find a method to get around  with it do you think ivy has a way to handle it

**Comment by faseehahmed26** [comment]
My test function and function are done just I am unable to find a way to handle this.

File: ivy/functional/frontends/torch/utilities.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+# global
+import torch
+
+# local
+import ivy
+
+
+def can_cast(from_, to):
+    res = 0
+    native_dtype_dict = {
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/torch/utilities.py:10]: There is an ongoing task for type promotion on the PyTorch frontend I think it is better that you can wait for it to be done then you can go on with the task.

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Well, I see that now there is no problem in making the first argument `from_` and I think the issue with the current implementation is that Ivy torch's frontend doesn't currently have type promotion that is being implemented now. We can put this task on hold while waiting for it to be finished. Sorry for the inconvenience.

File: ivy/functional/frontends/torch/utilities.py (hunk 9 above)
Comment (faseehahmed26) [reply ivy/functional/frontends/torch/utilities.py:10]: Can I work on other issues if possible

Comment (hirwa-nshuti) [reply ivy/functional/frontends/torch/utilities.py:10]: Yeah feel free to work on other task we will complete this one once type promotion is supported on the frontend side

**Pull request closed without merging**